File: apps/desktop/src/services/evaluation.ts
import { invoke } from '@tauri-apps/api/core';

export type EvaluationSuite = {
  id: string;
  workflowId: string;
  name: string;
  description: string;
  caseCount: number;
  createdAt: string;
  updatedAt: string;
};

export type EvaluationCase = {
  id: string;
  suiteId: string;
  name: string;
  description: string;
  position: number;
  enabled: boolean;
  archived: boolean;
  targetAgentId?: string | null;
  input: unknown;
  expectation: unknown;
  fixture: unknown;
  createdAt: string;
  updatedAt: string;
};

export type EvaluationRun = {
  id: string;
  suiteId: string;
  workflowId: string;
  retryOfRunId?: string | null;
  status: string;
  totalCases: number;
  startedAt: string;
};

export type EvaluationRunDetail = EvaluationRun & {
  endedAt?: string | null;
  durationMs?: number | null;
  passedCases: number;
  failedCases: number;
  totalTokens?: number | null;
  estimatedCostMicrousd?: number | null;
  error?: string | null;
};
export type EvaluationVersionSummary = {
  releaseId?: string | null;
  releaseVersion: string;
  comparisonKey: string;
  runCount: number;
  totalCases: number;
  passedCases: number;
  totalDurationMs: number;
  estimatedCostMicrousd: number;
};
export type EvaluationVersionCaseDiff = { caseId: string; name: string; baselineVerdict?: string | null; candidateVerdict?: string | null; kind: 'added' | 'removed' | 'regressed' | 'fixed' | 'persistent_failure' };
export type EvaluationCriterionOutcome = { criterion: string; passed: boolean; score: number; threshold: number; expected: unknown; actual: unknown };
export type EvaluationVersionCriterionDiff = { key: string; baseline?: EvaluationCriterionOutcome | null; candidate?: EvaluationCriterionOutcome | null; kind: 'added' | 'removed' | 'regressed' | 'fixed' | 'persistent_failure' | 'persistent_pass' };
export type EvaluationVersionCaseCriterionComparison = { caseId: string; name: string; baselineVerdict?: string | null; candidateVerdict?: string | null; criteria: EvaluationVersionCriterionDiff[] };
export type EvaluationQualityGate = {
  requireEvaluation: boolean;
  minPassRate?: number | null;
  maxCostMicrousd?: number | null;
  maxDurationMs?: number | null;
  requiredSuiteIds: string[];
};

export type EvaluationCaseResult = {
  id: string;
  evaluationCaseId: string;
  workflowRunId?: string | null;
  executionStatus: 'queued' | 'running' | 'completed' | 'failed' | 'cancelled';
  verdict: 'pending' | 'passed' | 'failed' | 'error' | 'skipped';
  score?: number | null;
  actualOutput: unknown;
  criteriaResults: unknown;
  normalizedTrace: unknown;
  failureReason?: string | null;
  durationMs?: number | null;
  totalTokens?: number | null;
  estimatedCostMicrousd?: number | null;
};

export type EvaluationWorkflowSnapshot = {
  targetName: string;
  targetSnapshot: unknown;
  dsl: unknown;
  releaseId?: string;
  releaseVersion?: string;
};

export function listEvaluationSuites(workflowId: string) {
  return invoke<EvaluationSuite[]>('evaluation_suite_list', { workflowId });
}

export function createEvaluationSuite(request: {
  id: string;
  workflowId: string;
  name: string;
  description: string;
}) {
  return invoke<EvaluationSuite>('evaluation_suite_create', { request });
}

export function updateEvaluationSuite(request: {
  id: string;
  name: string;
  description: string;
}) {
  return invoke<EvaluationSuite>('evaluation_suite_update', { request });
}

export function deleteEvaluationSuite(id: string) {
  return invoke('evaluation_suite_delete', { id });
}

export function listEvaluationCases(suiteId: string, includeArchived = false) {
  return invoke<EvaluationCase[]>('evaluation_case_list', { suiteId, includeArchived });
}

export function createEvaluationCase(request: {
  id: string;
  suiteId: string;
  name: string;
  description: string;
  position: number;
  enabled: boolean;
  targetAgentId?: string;
  input: unknown;
  expectation: unknown;
  fixture: unknown;
}) {
  return invoke<EvaluationCase>('evaluation_case_create', { request });
}

export function updateEvaluationCase(request: {
  id: string;
  name: string;
  description: string;
  enabled: boolean;
  targetAgentId?: string;
  input: unknown;
  expectation: unknown;
  fixture: unknown;
}) {
  return invoke<EvaluationCase>('evaluation_case_update', { request });
}

export function deleteEvaluationCase(id: string) {
  return invoke('evaluation_case_delete', { id });
}
export function restoreEvaluationCase(id: string) { return invoke('evaluation_case_restore', { id }); }

export function reorderEvaluationCases(suiteId: string, ids: string[]) {
  return invoke('evaluation_case_reorder', { suiteId, ids });
}

export function createEvaluationRun(request: {
  id: string;
  suiteId: string;
  workflowId: string;
  workflowSnapshot: EvaluationWorkflowSnapshot;
}) {
  return invoke<EvaluationRun>('evaluation_run_create', { request });
}

export function retryFailedEvaluationCases(sourceRunId: string, id: string) {
  return invoke<EvaluationRun>('evaluation_run_retry_failed', { sourceRunId, id });
}

export function cancelEvaluationRun(evaluationRunId: string) {
  return invoke('evaluation_run_cancel', { evaluationRunId });
}

export function listEvaluationCaseResults(evaluationRunId: string) {
  return invoke<EvaluationCaseResult[]>('evaluation_run_case_results', {
    evaluationRunId,
  });
}

export function inspectEvaluationRun(evaluationRunId: string) {
  return invoke<EvaluationRunDetail>('evaluation_run_inspect', {
    evaluationRunId,
  });
}

export function listEvaluationRuns(suiteId: string) {
  return invoke<EvaluationRunDetail[]>('evaluation_run_list', { suiteId });
}

export function summarizeEvaluationVersions(suiteId: string) {
  return invoke<EvaluationVersionSummary[]>('evaluation_version_summary', { suiteId });
}
export function compareEvaluationVersions(suiteId: string, baseline: string, candidate: string) { return invoke<EvaluationVersionCaseDiff[]>('evaluation_version_compare', { suiteId, baseline, candidate }); }
export function compareEvaluationVersionCaseCriteria(suiteId: string, baseline: string, candidate: string, caseId: string) {
  return invoke<EvaluationVersionCaseCriterionComparison>('evaluation_version_case_criteria_compare', { suiteId, baseline, candidate, caseId });
}

export function latestEvaluationRunForWorkflow(workflowId: string) {
  return invoke<EvaluationRunDetail | null>('evaluation_workflow_latest_run', { workflowId });
}

export function latestEvaluationRunsForWorkflowSnapshot(
  workflowId: string,
  workflowSnapshot: EvaluationWorkflowSnapshot,
) {
  return invoke<EvaluationRunDetail[]>('evaluation_workflow_snapshot_runs', {
    workflowId,
    workflowSnapshot,
  });
}

export function getEvaluationQualityGate(workflowId: string) {
  return invoke<EvaluationQualityGate>('evaluation_quality_gate_get', { workflowId });
}

export function updateEvaluationQualityGate(workflowId: string, policy: EvaluationQualityGate) {
  return invoke('evaluation_quality_gate_update', { workflowId, policy });
}

export function recordEvaluationQualityGateOverride(request: {
  workflowId: string;
  releaseVersion: string;
  reason: string;
  gateSnapshot: unknown;
  evaluationSnapshot: unknown;
}) {
  return invoke('evaluation_quality_gate_record_override', { request });
}
export type QualityGateAudit = { id: string; releaseVersion: string; actor: string; reason: string; gateSnapshot: unknown; evaluationSnapshot: unknown; createdAt: string };
export function listEvaluationQualityGateAudits(workflowId: string) { return invoke<QualityGateAudit[]>('evaluation_quality_gate_audit_list', { workflowId }); }

File: apps/desktop/src-tauri/src/cmd/evaluation.rs
use crate::{
    cmd::{CmdResult, StringifyErr},
    module::evaluation::{
        ClaimedEvaluationCase, CreateEvaluationCase, CreateEvaluationRun, CreateEvaluationSuite,
        EvaluationCaseResultSummary, EvaluationCaseSummary, EvaluationQualityGate, EvaluationRunDetail, EvaluationRunSummary, EvaluationStore, EvaluationVersionCaseDiff, EvaluationVersionSummary, QualityGateAuditSummary, RecordQualityGateOverride,
        EvaluationSuiteSummary, UpdateEvaluationCase, UpdateEvaluationSuite,
    },
};

#[tauri::command]
pub async fn evaluation_suite_create(request: CreateEvaluationSuite) -> CmdResult<EvaluationSuiteSummary> {
    EvaluationStore::create_suite(request).await.stringify_err()
}

#[tauri::command]
pub async fn evaluation_suite_list(workflow_id: String) -> CmdResult<Vec<EvaluationSuiteSummary>> {
    EvaluationStore::list_suites(&workflow_id).await.stringify_err()
}

#[tauri::command]
pub async fn evaluation_suite_update(request: UpdateEvaluationSuite) -> CmdResult<EvaluationSuiteSummary> {
    EvaluationStore::update_suite(request).await.stringify_err()
}

#[tauri::command]
pub async fn evaluation_suite_delete(id: String) -> CmdResult {
    EvaluationStore::delete_suite(&id).await.stringify_err()
}

#[tauri::command]
pub async fn evaluation_case_create(request: CreateEvaluationCase) -> CmdResult<EvaluationCaseSummary> {
    EvaluationStore::create_case(request).await.stringify_err()
}

#[tauri::command]
pub async fn evaluation_case_list(suite_id: String, include_archived: Option<bool>) -> CmdResult<Vec<EvaluationCaseSummary>> {
    EvaluationStore::list_cases(&suite_id, include_archived.unwrap_or(false)).await.stringify_err()
}

#[tauri::command]
pub async fn evaluation_case_update(request: UpdateEvaluationCase) -> CmdResult<EvaluationCaseSummary> {
    EvaluationStore::update_case(request).await.stringify_err()
}

#[tauri::command]
pub async fn evaluation_case_delete(id: String) -> CmdResult {
    EvaluationStore::delete_case(&id).await.stringify_err()
}

#[tauri::command]
pub async fn evaluation_case_restore(id: String) -> CmdResult {
    EvaluationStore::restore_case(&id).await.stringify_err()
}

#[tauri::command]
pub async fn evaluation_case_reorder(suite_id: String, ids: Vec<String>) -> CmdResult {
    EvaluationStore::reorder_cases(&suite_id, &ids).await.stringify_err()
}

#[tauri::command]
pub async fn evaluation_run_create(request: CreateEvaluationRun) -> CmdResult<EvaluationRunSummary> {
    EvaluationStore::create_run(request).await.stringify_err()
}

#[tauri::command]
pub async fn evaluation_run_claim_next_case(evaluation_run_id: String) -> CmdResult<Option<ClaimedEvaluationCase>> {
    EvaluationStore::claim_next_case(&evaluation_run_id)
        .await
        .stringify_err()
}

#[tauri::command]
pub async fn evaluation_run_start_next_case(evaluation_run_id: String) -> CmdResult<Option<ClaimedEvaluationCase>> {
    EvaluationStore::start_next_case(&evaluation_run_id)
        .await
        .stringify_err()
}

#[tauri::command]
pub async fn evaluation_run_case_results(evaluation_run_id: String) -> CmdResult<Vec<EvaluationCaseResultSummary>> {
    EvaluationStore::list_case_results(&evaluation_run_id)
        .await
        .stringify_err()
}

#[tauri::command]
pub async fn evaluation_run_inspect(evaluation_run_id: String) -> CmdResult<EvaluationRunDetail> {
    EvaluationStore::inspect_run(&evaluation_run_id).await.stringify_err()
}

#[tauri::command]
pub async fn evaluation_run_list(suite_id: String) -> CmdResult<Vec<EvaluationRunDetail>> {
    EvaluationStore::list_runs(&suite_id).await.stringify_err()
}

#[tauri::command]
pub async fn evaluation_version_summary(suite_id: String) -> CmdResult<Vec<EvaluationVersionSummary>> {
    EvaluationStore::summarize_versions(&suite_id).await.stringify_err()
}
#[tauri::command]
pub async fn evaluation_version_compare(suite_id: String, baseline: String, candidate: String) -> CmdResult<Vec<EvaluationVersionCaseDiff>> {
    EvaluationStore::compare_versions(&suite_id, &baseline, &candidate).await.stringify_err()
}

#[tauri::command]
pub async fn evaluation_workflow_latest_run(workflow_id: String) -> CmdResult<Option<EvaluationRunDetail>> {
    EvaluationStore::latest_run_for_workflow(&workflow_id).await.stringify_err()
}

#[tauri::command]
pub async fn evaluation_quality_gate_get(workflow_id: String) -> CmdResult<EvaluationQualityGate> {
    EvaluationStore::get_quality_gate(&workflow_id).await.stringify_err()
}

#[tauri::command]
pub async fn evaluation_quality_gate_update(workflow_id: String, policy: EvaluationQualityGate) -> CmdResult {
    EvaluationStore::update_quality_gate(&workflow_id, policy).await.stringify_err()
}

#[tauri::command]
pub async fn evaluation_quality_gate_record_override(request: RecordQualityGateOverride) -> CmdResult {
    EvaluationStore::record_quality_gate_override(request).await.stringify_err()
}

#[tauri::command]
pub async fn evaluation_quality_gate_audit_list(workflow_id: String) -> CmdResult<Vec<QualityGateAuditSummary>> {
    EvaluationStore::list_quality_gate_audits(&workflow_id).await.stringify_err()
}
